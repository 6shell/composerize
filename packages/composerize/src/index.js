// @flow

import 'core-js/fn/object/entries';

import yaml from 'yaml';
import parser from 'yargs-parser';
import deepmerge from 'deepmerge';
import Composeverter from 'composeverter';

import { maybeGetComposeEntry, getComposeJson, fromEntries } from './logic';

export type RawValue = string | number | boolean | [string | number | boolean] | any;

export type ComposeVersion = 'latest' | 'v2x' | 'v3x';

yaml.scalarOptions.null.nullStr = '';

const getServiceName = (image: string): string => {
    if (image === null || image === '' || typeof image === 'undefined') return '!!!invalid!!!';
    let name = image.includes('/') ? image.split('/').slice(-1)[0] : image;
    name = name.includes(':') ? name.split(':')[0] : name;

    return name;
};

export type ComposeFile = { composeFile: any, ignoredOptionsComments: string };

const getComposeFileJson = (input: string, existingComposeFile: string): ComposeFile => {
    const formattedInput = input
        .replace(/(\s)+/g, ' ')
        .trim()
        .replace(/\s-p(\d)/g, ' -p $1')
        .replace(/\s\\\s/g, ' ');
    const formattedInputArgs = formattedInput.replace(/^docker (run|create|container run|service create)/, '');
    const parsedInput: {
        +_: Array<string>,
        +[flag: string]: RawValue,
    } = parser(formattedInputArgs, {
        configuration: {
            'halt-at-non-option': true,
            'camel-case-expansion': false,
            'boolean-negation': false,
        },
        boolean: [
            'i',
            'interactive',
            't',
            'tty',
            'd',
            'detach',
            'rm',
            'init',
            'help',
            'privileged',
            'P',
            'publish-all',
            'no-healthcheck',
            'oom-kill-disable',
            'read-only',
            'sig-proxy',
        ],
    });
    const { _: command, ...params } = parsedInput;

    // The service object that we'll update
    let service = {};

    // $FlowFixMe: may be do better
    const { net: netArg, network: networkArg } = params;
    const network = netArg || networkArg || 'default';

    const ignoredOptions = [];
    // Loop through the tokens and append to the service object
    Object.entries(params).forEach(([key, value]: [string, RawValue]) => {
        // https://github.com/facebook/flow/issues/2174
        // $FlowFixMe: Object.entries wipes out types ATOW
        const result = maybeGetComposeEntry(key, value);
        if (result) {
            const entries = Array.isArray(result) ? result : [result];
            entries.forEach(entry => {
                // Store whatever the next entry will be
                const json = getComposeJson(entry, network);
                service = deepmerge(service, json);
            });
        } else {
            const dash = key.length === 1 ? '-' : '--';
            const valueString = value === true ? '' : `=${value.toString()}`;
            ignoredOptions.push(`# ${dash}${key}${valueString}`);
        }
    });

    const image = command[0];
    service.image = image;
    if (command.length > 1) {
        let argStart = 1;
        const commandArgsArray = [];
        while (argStart < command.length) {
            commandArgsArray.push(command[argStart]);
            argStart += 1;
        }
        service.command = commandArgsArray.join(' ');
    }

    const isNamedVolume = source => source && !source.includes('/') && !source.includes('\\') && !source.includes('$');
    const namedVolumes = [];
    if (service.volumes) {
        for (let volumeIndex = 0; volumeIndex < service.volumes.length; volumeIndex += 1) {
            let source;
            if (typeof service.volumes[volumeIndex] === 'string') {
                const volumeName = service.volumes[volumeIndex].split(':')[0];
                source = volumeName;
            } else {
                const volumeSource = service.volumes[volumeIndex].source;
                source = volumeSource;
            }
            if (isNamedVolume(source)) {
                namedVolumes.push([source, {}]);
            }
        }
    }

    const namedNetworks = [];
    if (service.networks) {
        if (Array.isArray(service.networks)) {
            for (let networkIndex = 0; networkIndex < service.networks.length; networkIndex += 1) {
                namedNetworks.push([
                    service.networks[networkIndex],
                    { external: { name: service.networks[networkIndex] } },
                ]);
            }
        } else {
            Object.keys(service.networks).forEach(serviceNetworkName => {
                // TODO: supposed to be done by babel : if (service.networks.hasOwnProperty(network))
                namedNetworks.push([serviceNetworkName, { external: { name: serviceNetworkName } }]);
            });
        }
    }

    const serviceName = getServiceName(image);

    // Outer template
    let result;
    const generatedCompose = {
        version: '3',
        services: {
            [serviceName]: service,
        },
    };
    const existingCompose = yaml.parse(existingComposeFile ?? '') ?? {};
    result = deepmerge(existingCompose, generatedCompose);
    if (namedNetworks.length > 0) {
        const networks = { networks: fromEntries(namedNetworks) };
        result = deepmerge(result, networks);
    }
    if (namedVolumes.length > 0) {
        const volumes = { volumes: fromEntries(namedVolumes) };
        result = deepmerge(result, volumes);
    }

    let ignoredOptionsComments = '';
    if (ignoredOptions.length > 0)
        ignoredOptionsComments = `# ignored options for '${serviceName}'\n${ignoredOptions.join('\n')}\n`;

    return ({
        composeFile: result,
        ignoredOptionsComments,
    }: ComposeFile);
};

export default (input: string, existingComposeFile: string = '', composeVersion: string = 'latest'): ?string => {
    const globalIgnoredOptionsComments = [];
    let result = {};
    const dockerCommands = input.split(/^\s*docker\s/gm);
    Object.values(dockerCommands).forEach(dockerCommand => {
        const command = String(dockerCommand);
        if (!command) return;
        if (!command.match(/^\s*(run|create|container run|service create)/)) {
            globalIgnoredOptionsComments.push(`# ignored : docker ${command}\n`);
            return;
        }
        const { composeFile, ignoredOptionsComments } = getComposeFileJson(`docker ${command}`, existingComposeFile);
        if (ignoredOptionsComments) globalIgnoredOptionsComments.push(ignoredOptionsComments);

        result = deepmerge(result, composeFile);
    });
    if (!result.services)
        throw new SyntaxError('must have at least a valid docker run/create/service create/container run command');

    let finalComposeYaml = yaml.stringify(result, { indent: 4, simpleKeys: true }).trim();
    if (composeVersion === 'v2x') finalComposeYaml = Composeverter.migrateFromV3xToV2x(finalComposeYaml);
    else if (composeVersion === 'latest') finalComposeYaml = Composeverter.migrateToCommonSpec(finalComposeYaml);
    else if (composeVersion !== 'v3x') throw new Error(`Unknown ComposeVersion '${composeVersion}'`);

    return globalIgnoredOptionsComments.join('\n') + finalComposeYaml;
};
