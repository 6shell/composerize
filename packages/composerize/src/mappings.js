// @flow

// Define the "types" of data a docker cli flag can represent in yaml.
export type ArgType =
    // Used for lists of things
    // e.g. --device (https://docs.docker.com/compose/compose-file/#devices)
    | 'Array'

    // Used to store a mapping of one key to one value
    // e.g. --log-driver (https://docs.docker.com/compose/compose-file/#logging)
    | 'KeyValue'

    // Used to store a "limits" value of the input format: <type>=<soft limit>[:<hard limit>]
    // e.g. --ulimit
    // @see https://docs.docker.com/compose/compose-file/#ulimits
    // @see https://docs.docker.com/engine/reference/commandline/run/#set-ulimits-in-container---ulimit
    | 'Ulimits'

    // Used to store a boolean value for an option
    // e.g. --privileged (https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir)
    | 'Switch'

    // Used to store an arbitrary text value for an option
    | 'Value'
    | 'IntValue'
    | 'FloatValue'
    | 'Networks'
    | 'MapArray'
    | 'Map'
    | 'Gpus';

// Type to represent the structure of the docker compose mapping
export type Mapping = {
    type: ArgType,
    path: string,
};

// Type to represent a compose file entry
export type ArrayComposeEntry = {
    path: string,
    value: [string],
};

export type KVComposeEntry = {
    path: string,
    value: {
        [string]: string | number | any,
    },
};

export type SwitchComposeEntry = {
    path: string,
    value: boolean,
};

export type ValueComposeEntry = {
    path: string,
    value: string | number | any,
};

export type IgnoreComposeEntry = {
    path: null,
    value: null,
};

export type ComposeEntry =
    | ArrayComposeEntry
    | KVComposeEntry
    | SwitchComposeEntry
    | ValueComposeEntry
    | IgnoreComposeEntry;

export const getMapping = (type: ArgType, path: string): Mapping => ({
    type,
    path,
});

// docker cli -> docker-compose options
export const MAPPINGS: { [string]: Mapping } = {
    'add-host': getMapping('Array', 'extra_hosts'),
    cap_add: getMapping('Array', 'cap_add'),
    cap_drop: getMapping('Array', 'cap_drop'),
    'cgroup-parent': getMapping('Value', 'cgroup_parent'),
    cgroupns: getMapping('Value', 'cgroup'),
    detached: getMapping('Switch', ''),
    device: getMapping('Array', 'devices'),
    dns: getMapping('Array', 'dns'),
    dns_search: getMapping('Array', 'dns_search'),
    env_file: getMapping('Array', 'env_file'),
    domainname: getMapping('Value', 'domainname'),
    expose: getMapping('Array', 'expose'),
    gpus: getMapping('Gpus', 'deploy'),
    'health-cmd': getMapping('Value', 'healthcheck/test'),
    'health-interval': getMapping('Value', 'healthcheck/interval'),
    'health-retries': getMapping('Value', 'healthcheck/retries'),
    'health-start-period': getMapping('Value', 'healthcheck/start_period'),
    'health-timeout': getMapping('Value', 'healthcheck/timeout'),
    hostname: getMapping('Value', 'hostname'),
    interactive: getMapping('Switch', 'stdin_open'),
    label: getMapping('Array', 'labels'),
    link: getMapping('Array', 'links'),
    'log-driver': getMapping('Array', 'logging/driver'),
    'log-opt': getMapping('KeyValue', 'logging/options'),
    entrypoint: getMapping('Array', 'entrypoint'),
    env: getMapping('Array', 'environment'),
    mount: getMapping('MapArray', 'volumes'),
    name: getMapping('Value', 'container_name'),
    net: getMapping('Networks', 'network_mode'), // alias for network
    'network-alias': getMapping('Array', 'networks/¤network¤/aliases'),
    network: getMapping('Networks', 'network_mode'),
    'no-healthcheck': getMapping('Switch', 'healthcheck/disable'),
    pid: getMapping('Value', 'pid'),
    privileged: getMapping('Switch', 'privileged'),
    publish: getMapping('Array', 'ports'),
    'read-only': getMapping('Switch', 'read_only'),
    restart: getMapping('Value', 'restart'),
    rm: getMapping('Switch', ''),
    tmpfs: getMapping('Value', 'tmpfs'),
    tty: getMapping('Switch', 'tty'),
    ulimit: getMapping('Ulimits', 'ulimits'),
    user: getMapping('Value', 'user'),
    volume: getMapping('Array', 'volumes'),
};

// Add flag mappings
MAPPINGS.v = MAPPINGS.volume;
MAPPINGS.p = MAPPINGS.publish;
MAPPINGS.e = MAPPINGS.env;
MAPPINGS.l = MAPPINGS.label;
MAPPINGS.h = MAPPINGS.hostname;
MAPPINGS.u = MAPPINGS.user;
MAPPINGS.t = MAPPINGS.tty;
MAPPINGS.i = MAPPINGS.interactive;
MAPPINGS.d = MAPPINGS.detached;
