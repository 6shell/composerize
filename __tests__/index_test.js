import Composerize from "../src";
import { maybeGetComposeEntry, getComposeJson } from "../src/logic";

// describe("maybeGetComposeEntry", () => {
//   it("works", () => {
//     // test '--device foo'
//     expect(maybeGetComposeEntry(["--device", "foo"], 0)).toMatchObject({
//       path: "devices",
//       value: ["foo"],
//     });
//   });
// });

// describe("getComposeJson", () => {
//   it("works", () => {
//     expect(
//       getComposeJson({
//         path: "devices",
//         value: ["foo"],
//       }),
//     ).toMatchObject({
//       devices: ["foo"],
//     });
//   });
// });

describe("snapshots", () => {
  const CMDS = [
    "docker run -p 80:80 -v /var/run/docker.sock:/tmp/docker.sock:ro --restart always --log-opt max-size=1g nginx",

    // test spacing
    " docker   run -p 80:80  -v /var/run/docker.sock:/tmp/docker.sock:ro --restart always --log-opt max-size=1g nginx    ",

    // test multiple args (https://github.com/magicmark/composerize/issues/9)
    'docker run -t --name="youtrack" -v /data/youtrack/data/:/opt/youtrack/data/ -v /data/youtrack/backup/:/opt/youtrack/backup/ -p 80:80 -p 3232:22351 uniplug/youtrack',

    // testing parsing of quotes (https://github.com/magicmark/composerize/issues/10)
    'docker run --name="foobar" nginx',

    "docker run -p 80:80 --name webserver nginx:latest",

    "docker run -p 80:80 foobar/baz:latest",
  ];

  it("match snapshots", () => {
    CMDS.forEach(cmd => {
      expect(Composerize(cmd)).toMatchSnapshot();
    });
  });
});
