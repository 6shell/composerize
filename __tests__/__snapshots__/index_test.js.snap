// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots match snapshots 1`] = `
"version: 3
services:
    nginx:
        ports:
            - '80:80'
        volumes:
            - '/var/run/docker.sock:/tmp/docker.sock:ro'
        restart: always
        logging:
            options:
                max-size: 1g
        image: nginx"
`;

exports[`snapshots match snapshots 2`] = `
"version: 3
services:
    nginx:
        ports:
            - '80:80'
        volumes:
            - '/var/run/docker.sock:/tmp/docker.sock:ro'
        restart: always
        logging:
            options:
                max-size: 1g
        image: nginx"
`;

exports[`snapshots match snapshots 3`] = `
"version: 3
services:
    youtrack:
        container_name: youtrack
        volumes:
            - '/data/youtrack/data/:/opt/youtrack/data/'
            - '/data/youtrack/backup/:/opt/youtrack/backup/'
        ports:
            - '80:80'
            - '3232:22351'
        image: uniplug/youtrack"
`;

exports[`snapshots match snapshots 4`] = `
"version: 3
services:
    nginx:
        container_name: foobar
        image: nginx"
`;

exports[`snapshots match snapshots 5`] = `
"version: 3
services:
    nginx:
        ports:
            - '80:80'
        container_name: webserver
        image: 'nginx:latest'"
`;

exports[`snapshots match snapshots 6`] = `
"version: 3
services:
    baz:
        ports:
            - '80:80'
        image: 'foobar/baz:latest'"
`;
